### PR TITLE
Fixed asan leak issues with RegexContext

### PR DESCRIPTION
**Pull request by bryancall**

### CODE REVIEW

File: src/tsutil/Regex.cc [hunk 1]
```diff
@@ -105,12 +120,20 @@ class RegexContext
 thread_local RegexContext *RegexContext::_regex_context = nullptr;
 
 //----------------------------------------------------------------------------
-namespace
+
+RegexContextCleanup::~RegexContextCleanup()
 {
```
Comment (moonchen) [inline src/tsutil/Regex.cc:125]: I think you need to acquire `_mutex` here, since it's modifying `_contexts`, right?

**Comment by ywkaras** [comment]
If you don't copy in these comments, https://github.com/apache/trafficserver/pull/11164/files#diff-8b89b132c9c9eb8fa6f58b050dd7bd3c47d55fd217d00239f4e13e21b270d679R161 , somebody may say "oh, I can just make these thread_local!" six months from now.

File: src/tsutil/Regex.cc (hunk 1 above)
Comment (bryancall) [reply src/tsutil/Regex.cc:125]: The mutex is not really needed, since it is called atexit, but I added it anyways.

**Comment by ywkaras** [comment]
Unfortunately, there is a race condition, if an object with static storage uses a DbgCtl object for the first time in its destructor.  (That requires checking if its tag matches the enabled tag regex.)  I think @cmcfarlen (persumably working in league with Satan) was able to generate a scenario like that.

**Comment by ywkaras** [comment]
Brian, were you going to change this so that all match requests will return false after the cleanup object is destroyed?

**Review by ywkaras — CHANGES_REQUESTED** [review]
Brian, were you going to change this so that all match requests will return false after the cleanup object is destroyed?

**Review by cmcfarlen — DISMISSED** [review]

File: src/tsutil/Regex.cc [hunk 2]
```diff
@@ -265,11 +305,19 @@ Regex::exec(std::string_view subject) const
...
+  // check if there is a compiled regex
   if (_code == nullptr) {
     return 0;
   }
+
+  // get the RegexContext instance - should only be null when shutting down
+  RegexContext *regex_context = RegexContext::get_instance();
+  if (regex_context == nullptr) {
+    return false;
```
Comment (ywkaras) [inline src/tsutil/Regex.cc:316]: Seems like this should return 0.

Comment (bryancall) [reply src/tsutil/Regex.cc]: Fixed, copy-and-paste issue.

**Review by cmcfarlen — APPROVED** [review]

**Review by ywkaras — APPROVED** [review]

File: src/tsutil/DbgCtl.cc [hunk 3]
```diff
@@ -150,14 +150,28 @@ DbgCtl::_new_reference(char const *tag)
...
+  // 1.  Thread 1 calls a DbgCtl constructor, which locks the registry mutex, but then is suspended.
+  // 2.  Thread 2 calls dlopen() for a plugin, locking the runtime mutex.  It then executes the constructor for a
+  //     statically allocated DbgCtl object, which blocks on locking the registry mutex.
+  // 3.  Thread 1 resumes, and calls member functions of the derived class of DebugInterface.  If this causes the
+  //     the construction of a thread_local variable with a non-trivial destructor, _cxa_thread_atexit() will be called
+  //     to set up a call of the variable's destructor at thread exit.  The call to _cxa_thread_atexit() will block on
+  //     the runtime mutex (held by Thread 2).  So Thread 1 holds the registry mutex and is blocked waiting for the
+  //     runtime mutex.  And Thread 2 holds the runtime mutex and is blocked waiting for the registry mutex.  Deadlock.
+  //
+  // This deadlock is avoided by having the thread_local variable register its destruction in a non-thread_local class.
```
Comment (ywkaras) [inline src/tsutil/DbgCtl.cc:174]: ??? I thought this was already present.

**Comment by bryancall** [comment]
[approve ci autest]

**Pull request merged**